File: 0101-symmetric-tree/0101-symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool f(TreeNode* nodeL, TreeNode* nodeR) {
        if(nodeL == nullptr and nodeR != nullptr) return false;
        if(nodeL != nullptr and nodeR == nullptr) return false;
        if(nodeL == nullptr and nodeR == nullptr) return true;
        if(nodeL->val != nodeR->val) return false;
        bool r1 = f(nodeL->left, nodeR->right);
        bool r2 = f(nodeL->right, nodeR->left);
        return r1 and r2;
    }
    bool isSymmetric(TreeNode* root) {
        return f(root->left, root->right);
    }
};

File: 0101-symmetric-tree/0101-symmetric-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        return f(root.left, root.right);
    }
    private boolean f(TreeNode nodeL, TreeNode nodeR) {
        if(nodeL == null && nodeR == null) return true;
        if(nodeL == null && nodeR != null) return false;
        if(nodeL != null && nodeR == null) return false;
        
        if(nodeL.val != nodeR.val) return false;
        
        boolean r1 = f(nodeL.left, nodeR.right);
        boolean r2 = f(nodeL.right, nodeR.left);
        
        return r1 && r2;
    }
}